libyul/optimiser/LabelIDDispenser.cpp

```diff
@@ -0,0 +1,168 @@
+
+#include <libyul/optimiser/LabelIDDispenser.h>
+
+#include <libyul/optimiser/OptimizerUtilities.h>
+
+#include <fmt/compile.h>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/iota.hpp>
+
+using namespace solidity::yul;
+
+namespace
+{
+bool isInvalidLabel(
+	std::string_view const _label,
+	std::set<std::string, std::less<>> const& _reservedLabels,
+	Dialect const& _dialect
+)
+{
+	return isRestrictedIdentifier(_dialect, _label) || _reservedLabels.contains(_label);
+}
+}
+
+LabelIDDispenser::LabelIDDispenser(ASTLabelRegistry const& _labels, std::set<std::string> const& _reservedLabels):
+	m_labels(_labels),
+	m_reservedLabels(_reservedLabels.begin(), _reservedLabels.end())
+{}
+
+LabelIDDispenser::LabelID LabelIDDispenser::newID(LabelID const _parent)
+{
+	auto const parentLabelID = resolveParentLabelID(_parent);
+	yulAssert(!m_labels.ghost(parentLabelID), "Use newGhost to add new ghosts.");
+	m_idToLabelMapping.push_back(parentLabelID);
+	return m_idToLabelMapping.size() + m_labels.maxID();
+}
+
+LabelIDDispenser::LabelID LabelIDDispenser::newGhost()
+{
+	m_idToLabelMapping.push_back(ASTLabelRegistry::ghostLabelIndex());
+	return m_idToLabelMapping.size() + m_labels.maxID();
+}
+
+LabelIDDispenser::LabelID LabelIDDispenser::resolveParentLabelID(LabelID _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size() + m_labels.maxID() + 1, "ID exceeds bounds.");
+	// bigger than maxID means that the input label id that was spawned by this dispenser
+	if (_id > m_labels.maxID())
+		_id = m_idToLabelMapping[_id - m_labels.maxID() - 1];
+	yulAssert(
+		_id <= m_labels.maxID() && !m_labels.unused(_id),
+		"We can have at most one level of indirection and the derived-from label cannot be unused."
+	);
+	return _id;
+}
+
+bool LabelIDDispenser::ghost(LabelID const _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size() + m_labels.maxID() + 1, "ID exceeds bounds.");
+	if (_id > m_labels.maxID())
+		return m_idToLabelMapping[_id - m_labels.maxID() - 1] == ASTLabelRegistry::ghostLabelIndex();
+
+	return m_labels.ghost(_id);
+}
+
+ASTLabelRegistry LabelIDDispenser::generateNewLabels(Dialect const& _dialect) const
+{
+	auto const usedIDs =
+		ranges::views::iota(static_cast<size_t>(1), m_idToLabelMapping.size() + m_labels.maxID() + 1) |
+		ranges::to<std::set>;
+	return generateNewLabels(usedIDs, _dialect);
+}
+
+ASTLabelRegistry LabelIDDispenser::generateNewLabels(std::set<LabelID> const& _usedIDs, Dialect const& _dialect) const
+{
+	if (_usedIDs.empty())
+		return {};
+
+	auto const& originalLabels = m_labels.labels();
+
+	std::vector<uint8_t> reusedLabels (originalLabels.size());
+	// this means that everything that is derived from empty needs to be generated
+	reusedLabels[0] = true;
+
+	// start with the empty label
+	std::vector<std::string> labels{""};
+	labels.reserve(originalLabels.size()+1);
+
+	// 0 maps to ""
+	std::vector<size_t> idToLabelMap = {0};
+	if (!_usedIDs.empty())
```

Comment:
```suggestion
	yulAssert (!_usedIDs.empty())
```

Reply:
ah yes, now that the case of `used_IDs` empty is caught further up, absolutely makes sense to make this into an assert and also do the `idToLabelMap` in one go via ctor call!